app/src/components/templateEditor/language/languageUtils.ts

```diff
@@ -0,0 +1,86 @@
+import { LRParser } from "@lezer/lr";
+
+/**
+ * Extracts all variables from a templated string.
+ *
+ * @param parser - The parser for the templating language.
+ *  The parser should be a language parser that emits Variable nodes.
+ * @param text - The text to extract variables from.
+ *
+ * @returns An array of variable names.
+ */
+export const extractVariables = (parser: LRParser, text: string) => {
+  const tree = parser.parse(text);
+  const variables: string[] = [];
+  const cur = tree.cursor();
+  do {
+    if (cur.name === "Variable") {
+      variables.push(text.slice(cur.node.from, cur.node.to));
+    }
+  } while (cur.next());
+  return variables;
+};
+
+/**
+ * Formats a templated string with the given variables.
+ *
+ * The parser should be a language parser that emits Variable nodes as children of some parent node.
+ */
+export const format = ({
+  parser,
+  text,
+  variables,
+  postFormat,
+}: {
+  /**
+   * The parser for the templating language.
+   *
+   * Should be MustacheLikeTemplatingLanguage or FStringTemplatingLanguage.
+   *
+   * format assumes that the language produces a structure where Variable nodes
+   * are children of some parent node, in this case Template.
+   */
+  parser: LRParser;
+  /**
+   * The text to format.
+   */
+  text: string;
+  /**
+   * A mapping of variable names to their values.
+   *
+   * If a variable is not found in this object, it will be left as is.
+   */
+  variables: Record<string, string | number | boolean>;
```

Comment:
nit: your check below avoids this, but sometimes prefer writing records as Record<string, string | number | undefined> just because it's a little more honest and catches unset keys

Reply:
in my opinion we should just enable `noUncheckedIndexedAccess` in our tsconfig if we are concerned about unset keys

Reply:
Added